monkey/infection_monkey/network/ping_scanner.py

```diff
@@ -1,79 +1,83 @@
-            errors="backslashreplace",
-        )
-
-        logger.debug(f"Retrieving ping command output using {encoding} encoding")
-        output = " ".join(sub_proc.communicate())
-        regex_result = self._ttl_regex.search(output)
-        if regex_result:
-            try:
-                ttl = int(regex_result.group(0))
-                if ttl <= LINUX_TTL:
-                    host.os["type"] = "linux"
-                else:  # as far we we know, could also be OSX/BSD but lets handle that when it
-                    # comes up.
-                    host.os["type"] = "windows"
-
-                host.icmp = True
-
-                return True
-            except Exception as exc:
-                logger.debug("Error parsing ping fingerprint: %s", exc)
-
-        return False
-
-    def _build_ping_command(self, ip_addr):
-        return ["ping", PING_COUNT_FLAG, "1", PING_TIMEOUT_FLAG, str(self._timeout), ip_addr]
+def ping(host: str, timeout: float) -> PingScanData:
+    if "win32" == sys.platform:
+        timeout = math.floor(timeout * 1000)
+
+    ping_command_output = _run_ping_command(host, timeout)
+
+    ping_scan_data = _process_ping_command_output(ping_command_output)
+    logger.debug(f"{host} - {ping_scan_data}")
+
+    return ping_scan_data
+
+
+def _run_ping_command(host: str, timeout: float) -> str:
+    ping_cmd = _build_ping_command(host, timeout)
+    logger.debug(f"Running ping command: {' '.join(ping_cmd)}")
+
+    # If stdout is not connected to a terminal (i.e. redirected to a pipe or file), the result
+    # of os.device_encoding(1) will be None. Setting errors="backslashreplace" prevents a crash
+    # in this case. See #1175 and #1403 for more information.
+    encoding = os.device_encoding(1)
+    sub_proc = subprocess.Popen(
+        ping_cmd,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        text=True,
+        encoding=encoding,
+        errors="backslashreplace",
+    )
+
+    logger.debug(f"Retrieving ping command output using {encoding} encoding")
+    # The underlying ping command should timeout within the specified timeout. Setting the timeout
+    # parameter on communicate() is a failsafe mechanism for ensuring this does not block
+    # indefinitely. It is set to a multiplier of timeout to prevent the communicate() call from
+    # returning prematurely.
+
+    try:
+        output = " ".join(sub_proc.communicate(timeout=(timeout * 10)))
+        logger.debug(output)
+    except subprocess.TimeoutExpired as te:
+        logger.error(te)
+        return ""
+
+    return output
+
+
+def _process_ping_command_output(ping_command_output: str) -> PingScanData:
+    regex_result = TTL_REGEX.search(ping_command_output)
+    if not regex_result:
+        return PingScanData(False, None)
+
+    # It should be impossible for this next line to raise any errors, since the TTL_REGEX won't
+    # match at all if the group isn't found or the contents of the group are not only digits.
```

Comment:
unnecessary comment?

Reply:
Maybe. In the past, there was a try/except around this, so this comment serves two purposes:
1. Indicate to reviewers that they should double-check my logic.
2. If any bugs arise in the future, this comment may help someone find the bug faster.

I'm open to removing it, though, if everyone agrees a try/except isn't necessary here.